File: entity-meta-generator-lib/src/main/java/com/loudsight/meta/serialization/transform/SetEntityTransform.java
package com.loudsight.meta.serialization.transform;

import com.loudsight.meta.serialization.EntityType;

import java.util.HashSet;
import java.util.Set;
import java.util.function.Function;

public class SetEntityTransform extends CollectionEntityTransform<Set<?>> {

    private static class SetEntityTransformHolder {
        private static final SetEntityTransform INSTANCE = new SetEntityTransform();
    }
    // global access point
    public static SetEntityTransform getInstance() {
        return SetEntityTransformHolder.INSTANCE;
    }


    private SetEntityTransform() {
        super(EntityType.SET, (Class<Set<?>>)(Object)Set.class);
    }

    @Override public Set<?> newCollection(int size, Function<Integer, Object> init) {
        var set = new HashSet<>(size);
        for (int i  = 0; i  < size; i ++) {
            set.add(init.apply(i));
        }
        return set;
    }
}

File: entity-meta-generator-lib/src/main/kotlin/com/loudsight/meta/serialization/transform/SetEntityTransform.kt
package com.loudsight.meta.serialization.transform

import com.loudsight.meta.serialization.EntityType

object SetEntityTransform : CollectionEntityTransform<Set<*>>(EntityType.SET, Set::class) {

    override fun newCollection(size: Int, init: (index: Int) -> Any?): Set<*> {
        val set = HashSet<Any>(size)
        for (i in 0 until size) {
            set.add(init.invoke(i) as Any)
        }
        return set
    }

    override fun canTransform(entity: Any): Boolean {
        return entity is Set<*>
    }
}